src/lib/optimizer/strategy/between_composition_rule.cpp

```diff
@@ -0,0 +1,265 @@
+#include "between_composition_rule.hpp"
+
+#include <algorithm>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "constant_mappings.hpp"
+#include "logical_query_plan/abstract_lqp_node.hpp"
+#include "logical_query_plan/lqp_utils.hpp"
+#include "logical_query_plan/predicate_node.hpp"
+#include "statistics/table_statistics.hpp"
+#include "utils/assert.hpp"
+
+using namespace opossum::expression_functional;  // NOLINT
+
+namespace opossum {
+
+std::string BetweenCompositionRule::name() const { return "Between Composition Rule"; }
+
+/**
+ * _get_boundary takes a BinaryPredicateExpression and the corresponding PredicateNode
+ * as its input and returns a normalized ColumnBoundary. This function checks where the
+ * LQPColumnExpression and where the ValueExpression is stored in the BinaryPredicateExpression.
+ * The expressions are transferred to a normalized ColumnBoundary format
+ * and are labled with a ColumnBoundaryType, that depends on their positions and the predicate condition
+ * of the BinaryPredicateExpression
+ *
+ **/
+const ColumnBoundary BetweenCompositionRule::_get_boundary(const std::shared_ptr<BinaryPredicateExpression>& expression,
+                                                           const std::shared_ptr<PredicateNode>& node) const {
+  auto type = ColumnBoundaryType::None;
+  auto column_expression = std::dynamic_pointer_cast<LQPColumnExpression>(expression->left_operand());
+  auto value_expression = std::dynamic_pointer_cast<ValueExpression>(expression->right_operand());
+
+  // Case: "ColumnExpression [CONDITION] ValueExpression" will be checked
+  // Boundary type will be set according to this order
+  if (column_expression != nullptr && value_expression != nullptr) {
+    switch (expression->predicate_condition) {
+      case PredicateCondition::LessThanEquals:
+        type = ColumnBoundaryType::UpperBoundaryInclusive;
+        break;
+      case PredicateCondition::GreaterThanEquals:
+        type = ColumnBoundaryType::LowerBoundaryInclusive;
+        break;
+      case PredicateCondition::LessThan:
+        type = ColumnBoundaryType::UpperBoundaryExclusive;
+        break;
+      case PredicateCondition::GreaterThan:
+        type = ColumnBoundaryType::LowerBoundaryExclusive;
+        break;
+      default:
+        break;
+    }
+  } else {
+    value_expression = std::dynamic_pointer_cast<ValueExpression>(expression->left_operand());
+    column_expression = std::dynamic_pointer_cast<LQPColumnExpression>(expression->right_operand());
+
+    // Case: "ValueExpression [CONDITION] ColumnExpression" will be checked
+    // Boundary type will be set according to this order
+    if (value_expression != nullptr && column_expression != nullptr) {
```

Comment:
So the rule does NOT work for things like `a >= b AND a <= c`? I think this would be more in its scope than pruning expression (referring to the "find optimal between" thing). If you have to sacrifice one thing to make the other work, pls sacrifice the latter instead of the first.